catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultPaginator.java

```diff
@@ -0,0 +1,156 @@
+
+    private final CatalogFramework catalogFramework;
+
+    private final QueryRequestImpl queryRequestCopy;
+
+    private int currentIndex;
+
+    private int pageSize;
+
+    //Set to true when a query returns zero results.
+    private boolean exhausted;
+
+    private BlockingQueue<Result> buffer;
+
+    public QueryResultPaginator(CatalogFramework catalogFramework, QueryRequestImpl queryRequest) {
+        Validate.notNull(catalogFramework);
+        Validate.notNull(queryRequest);
+        this.catalogFramework = catalogFramework;
+        this.queryRequestCopy = clone(queryRequest);
+        pageSize = queryRequestCopy.getQuery()
+                .getPageSize();
+        currentIndex = queryRequest.getQuery()
+                .getStartIndex();
+        Validate.inclusiveBetween(1,
+                Integer.MAX_VALUE,
+                currentIndex,
+                "Query request start index must be greater than zero");
+        exhausted = false;
+        buffer = new LinkedBlockingDeque<>();
+    }
+
+    /**
+     * @return true if there are more results for the client
+     */
+    public boolean hasNext() {
+
+        if (exhausted && buffer.isEmpty()) {
+            return false;
+        }
+
+        if (!exhausted) {
+            fetchNext();
+        }
+
+        return !buffer.isEmpty();
+    }
+
+    /**
+     * Catalog Framework queried for results until current page is full of results
+     * or endIndex is reached.
+     *
+     * @return Collection of results
+     * @throws NoSuchElementException if the iteration has no more elements
+     */
+    public List<Result> next() throws NoSuchElementException {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        while (!exhausted && getBufferSize() < pageSize) {
+            fetchNext();
+        }
+
+        List<Result> copy = new ArrayList<>(pageSize);
+        buffer.drainTo(copy, pageSize);
+        return copy;
+    }
+
+    int getBufferSize() {
+        return buffer.size();
+    }
+
+    QueryRequestImpl clone(QueryRequestImpl queryRequest) {
+        QueryImpl query = (QueryImpl) queryRequest.getQuery();
+        return new QueryRequestImpl(new QueryImpl(query.getFilter(),
+                query.getStartIndex(),
+                query.getPageSize(),
+                query.getSortBy(),
+                query.requestsTotalResultsCount(),
+                query.getTimeoutMillis()));
+    }
+
+    int getCurrentIndex() {
+        return currentIndex;
+    }
+
+    /**
+     * Get the next batch of results.
+     * Assume the current index is pointed to the first result
+     * that has not been retrieved.
+     * Update the current index after the results are fetched.
+     * Add the fetched results to the buffer.
+     * If no results were returned, assume there are no more results possible.
+     */
+    private void fetchNext() {
+        ((QueryImpl) queryRequestCopy.getQuery()).setStartIndex(currentIndex);
```

Comment:
❗️ Again, I'm not clear how you can know with any certainty that this is a `QueryImpl`. Although, if your `clone/copy` method were to explicitly create a `QueryImpl` that might help.